packages/taiko-client/driver/preconf_blocks/server.go

```diff
@@ -310,6 +401,374 @@ func (s *PreconfBlockAPIServer) OnUnsafeL2Payload(
+		log.Debug(
+			"Ignoring OnUnsafeL2Request, not the current sequencer",
+			"peer", from,
+			"currOperator", s.lookahead.CurrOperator.Hex(),
+			"nextOperator", s.lookahead.NextOperator.Hex(),
+			"preconfOperatorAddress", s.preconfOperatorAddress.Hex(),
+		)
+
+		return nil
+	}
+
+	log.Info("OnUnsafeL2Request New block request from p2p network", "peer", from, "hash", hash.Hex())
+
+	block, err := s.rpc.L2.BlockByHash(ctx, hash)
+	if err != nil {
+		log.Warn("OnUnsafeL2Request Failed to fetch block by hash", "hash", hash.Hex(), "error", err)
+		return err
+	}
+
+	var u256 uint256.Int
+	if overflow := u256.SetFromBig(block.BaseFee()); overflow {
+		log.Warn(
+			"OnUnsafeL2Request Failed to convert base fee to uint256, skip propagating the preconfirmation block",
+			"baseFee", block.BaseFee,
+		)
+	} else {
+		txs, err := utils.EncodeAndCompressTxList(block.Transactions())
+		if err != nil {
+			return err
+		}
+
+		envelope := &eth.ExecutionPayloadEnvelope{
+			ExecutionPayload: &eth.ExecutionPayload{
+				BaseFeePerGas: eth.Uint256Quantity(u256),
+				ParentHash:    block.ParentHash(),
+				FeeRecipient:  block.Coinbase(),
+				ExtraData:     block.Extra(),
+				PrevRandao:    eth.Bytes32(block.MixDigest()),
+				BlockNumber:   eth.Uint64Quantity(block.NumberU64()),
+				GasLimit:      eth.Uint64Quantity(block.GasLimit()),
+				GasUsed:       eth.Uint64Quantity(block.GasUsed()),
+				Timestamp:     eth.Uint64Quantity(block.Time()),
+				BlockHash:     block.Hash(),
+				Transactions:  []eth.Data{hexutil.Bytes(txs)},
+			},
+		}
+
+		if err := s.ValidateExecutionPayload(envelope.ExecutionPayload); err != nil {
+			log.Warn(
+				"OnUnsafeL2Request Invalid preconfirmation block payload",
+				"peer", from,
+				"blockID", uint64(envelope.ExecutionPayload.BlockNumber),
+				"hash", envelope.ExecutionPayload.BlockHash.Hex(),
+				"parentHash", envelope.ExecutionPayload.ParentHash.Hex(),
+				"error", err,
+			)
+			return nil
+		}
+
+		log.Info("OnUnsafeL2Request publishing response",
+			"hash", hash.Hex(),
+			"blockID", block.NumberU64(),
+		)
+
+		if err := s.p2pNode.GossipOut().PublishL2RequestResponse(ctx, envelope, s.p2pSigner); err != nil {
+			log.Warn("OnUnsafeL2Request failed to publish",
+				"error", err,
+				"hash", hash.Hex(),
+				"blockID", uint64(envelope.ExecutionPayload.BlockNumber),
+			)
+		}
+	}
+
+	return nil
+}
+
+// OnUnsafeL2EndOfSequencingRequest implements the p2p.GossipIn interface.
+func (s *PreconfBlockAPIServer) OnUnsafeL2EndOfSequencingRequest(
+	ctx context.Context,
+	from peer.ID,
+	epoch uint64,
+) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	// Ignore the message if it is from the current P2P node.
+	if from != "" && s.p2pNode.Host().ID() == from {
+		log.Debug("Ignore the message from the current P2P node", "peer", from)
+
+		return nil
+	}
+
+	// only respond if you are the current sequencer, *not* the active sequencer in the slot.
+	if s.preconfOperatorAddress.Hex() != s.lookahead.CurrOperator.Hex() {
```

Comment:
If the sidecar just starts at its own slot, not during the last one's handover window, will there be an issue?

Reply:
Yes, we should actually let anyone respond with the current block and remove this check, but just make sure that if many people respond with the same block, that it's OK, and we also don't hit the limit in the optimism lib to ignore messages.